iwyu_ast_util.cc

```diff
@@ -556,24 +557,29 @@ class TypeEnumerator : public RecursiveASTVisitor<TypeEnumerator> {
 
   set<const Type*> Enumerate(const TemplateArgument& tpl_arg) {
     seen_types_.clear();
-    TraverseTemplateArgument(tpl_arg);
+    this->TraverseTemplateArgument(tpl_arg);
     return seen_types_;
   }
 
   // --- Methods on RecursiveASTVisitor
   bool TraverseType(QualType type) {
     if (type.isNull())
       return Base::TraverseType(type);
-    return TraverseTypeHelper(type);
+    return this->getDerived().TraverseTypeHelper(type);
   }
 
   bool TraverseTypeLoc(TypeLoc type_loc) {
     if (!type_loc)
       return Base::TraverseTypeLoc(type_loc);
-    return TraverseTypeHelper(type_loc.getType());
+    return this->getDerived().TraverseTypeHelper(type_loc.getType());
   }
 
- private:
+ protected:
+  set<const Type*> seen_types_;
+};
+
+class TemplateArgEnumerator : public TypeEnumerator<TemplateArgEnumerator> {
```

Comment:
I was surprised by this name, I would have expected this to cover all sorts of types that aren't necessarily template arguments?

This looks more to me like `SugaredTypeEnumerator`?

Reply:
This is just use case encoded in the name. But following that logic, `TypeEnumeratorWithoutSubstituted` should be named as `TypeEnumeratorToGetProvidedTypes` or something similar... On the one hand, it is not a good approach. On the other hand, these "enumerators" are inherently very specific.

Reply:
Aha, I didn't see it that way at all (use-case encoded in name). I was thinking it enumerated sugared types and returned a desugared set. But maybe `DesugaredTypeEnumerator` makes it clearer what it produces? Or I'm entirely confused.

Reply:
No, just recall: it is used to produce _sugared_ side of `resugar_map`. But it keeps only "interesting" sugar (typedefs, template specializations, see `desugared_until_alias_or_tpl`). If that sugar hides itself another complex type, the visitor traverses its components too.